File: eco-core/core-plugin/src/main/java/com/willfp/ecoarmor/util/DiscoverRecipeListener.java
package com.willfp.ecoarmor.util;

import com.willfp.eco.util.internal.PluginDependent;
import com.willfp.eco.util.plugin.AbstractEcoPlugin;
import org.bukkit.Bukkit;
import org.bukkit.NamespacedKey;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.inventory.ShapedRecipe;
import org.jetbrains.annotations.NotNull;

public class DiscoverRecipeListener extends PluginDependent implements Listener {
    /**
     * Register listener.
     *
     * @param plugin Talismans.
     */
    public DiscoverRecipeListener(@NotNull final AbstractEcoPlugin plugin) {
        super(plugin);
    }

    /**
     * Unlock all recipes on player join.
     *
     * @param event The event to listen for.
     */
    @EventHandler
    public void onJoin(@NotNull final PlayerJoinEvent event) {
        Player player = event.getPlayer();

        if (this.getPlugin().getConfigYml().getBool("discover-recipes")) {
            Bukkit.getServer().recipeIterator().forEachRemaining(recipe -> {
                if (recipe instanceof ShapedRecipe) {
                    NamespacedKey key = ((ShapedRecipe) recipe).getKey();
                    if (key.getNamespace().equals("ecoarmor")) {
                        if (!key.getKey().contains("displayed")) {
                            player.discoverRecipe(key);
                        }
                    }
                }
            });
        }
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/ecoarmor/util/DiscoverRecipeListener.kt
package com.willfp.ecoarmor.util

import com.willfp.ecoarmor.plugin
import org.bukkit.Bukkit
import org.bukkit.Keyed
import org.bukkit.NamespacedKey
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent

object DiscoverRecipeListener : Listener {
    private var cachedRecipeKeys = emptyList<NamespacedKey>()

    fun reloadRecipeCache() {
        val namespace = plugin.name.lowercase()
        cachedRecipeKeys = buildList {
            Bukkit.getServer().recipeIterator().forEachRemaining { recipe ->
                if (recipe is Keyed) {
                    val key = recipe.key
                    if (key.namespace == namespace && !key.key.contains("displayed")) {
                        add(key)
                    }
                }
            }
        }
    }

    @EventHandler
    fun onJoin(event: PlayerJoinEvent) {
        if (!plugin.configYml.getBool("discover-recipes")) {
            return
        }
        for (key in cachedRecipeKeys) {
            event.player.discoverRecipe(key)
        }
    }
}